protocol/src/protocol.callHierarchy.proposed.md

```diff
@@ -0,0 +1,148 @@
+
+#### Call Hierarchy
+
+The LSP provides retrieving the call hierachy information with the following request.
+
+_Client Capabilities_:
+
+```ts
+CallHierarchyClientCapabilities {
+    /**
+     * The text document client capabilities
+     */
+    textDocument?: {
+        /**
+         * Capabilities specific to the `textDocument/callHierarchy`
+         */
+        callHierarchy?: {
+            /**
+             * Whether implementation supports dynamic registration. If this is set to `true`
+             * the client supports the new `(TextDocumentRegistrationOptions & StaticRegistrationOptions)`
+             * return value for the corresponding server capability as well.
+             */
+            dynamicRegistration?: boolean;
+        };
+}
+```
+
+_Server Capabilities_:
+
+```ts
+CallHierarchyServerCapabilities {
+    /**
+     * The server provides Call Hierarchy support.
+     */
+    callHierarchyProvider?: boolean | (TextDocumentRegistrationOptions & StaticRegistrationOptions);
+}
+```
+
+##### Call Hierarchy Request
+
+_Request_:
+
+The `textDocument/callHierarchy` request is sent from the client to the server to request the call hierarchy for a symbol at the given text document position.
+
+Returns a call hierarchy item for the requested call direction.
+
+* method: ‘textDocument/callHierarchy'
+* params: `CallHierarchyParams` defined as follows:
+
+```ts
+export interface CallHierarchyParams extends TextDocumentPositionParams {
+    resolve?: number;
+    direction?: CallHierarchyDirection;
+}
+```
+
+_Response_:
+
+The server will send a `CallHierarchyItem` object containing the information about the targeted symbol. The item will be undefined, if no such symbol is found.
+
+The item is _unresolved_ if the lists of callers and callees are undefined. Unresolved items can be resolved via `callHierarchy/resolve` requests.
+
+The resolved item includes callers or callees again of type `CallHierarchyItem`. The caller/callee object provide the actual location of the call (`CallHierarchyItem.callLocation`).
+
+* result: `CallHierarchyItem` | `null` defined as follows:
+
+```ts
+export interface CallHierarchyItem {
+
+    /**
+     * The name of the symbol targeted by the call hierarchy request.
+     */
+    name: string;
+
+    /**
+     * More detail for this symbol, e.g the signature of a function.
+     */
+    detail?: string;
+
+    /**
+     * The kind of this symbol.
+     */
+    kind: SymbolKind;
+
+    /**
+     * `true` if the hierarchy item is deprecated. Otherwise, `false`. It is `false` by default.
+     */
+    deprecated?: boolean;
```

Comment:
Not sure that `deprecated` fits well into the call hierarchy domain. I understand the desire to represent if a symbol is deprecated but that's also true for completions or outline.  I think deprecated-knowledge deserves it's in own, independent API

Reply:
+1 for removig it.

Reply:
+1 for removing.

Reply:
Actually the LSP supports deprectated on `CompletionItem` and `DocumentSymbol`. @jrieken how would a independent API look like in your opinion. When LSP introduced it it seemed to be the most logical to add it to the items. The only thing I could think of having it as a tag now that we started to introduce this for diagnostics. But it would still go onto the items. Or?

Reply:
yeah, never thought much about it but I think I would have done something similar to diagnostics or decorations. However, if LSP already gone down this road for completions and outline that I think its fair to continue like that